added a comment that statuses lists are not being validated

SergeyKanzhelev opened the pull request:
#### What type of PR is this?

/kind cleanup
/kind documentation
/sig node
/priority backlog

#### What this PR does / why we need it:

Adding a note to API will help troubleshoot issues in future. Ideally we need to add this validation to API, but it is not a priority and will require a big effort of ensuring backward compatibility. Thus I suggest we are limiting change with the documentation update.

#### Which issue(s) this PR fixes:

Related to: https://github.com/kubernetes/kubernetes/issues/124915
Fixes: https://github.com/kubernetes/kubernetes/issues/120676

#### Special notes for your reviewer:

I also reviewed other instances where statuses collection is being iterated over. I didn't find anything dangerous there.

Please review commit-by-commit. It will be easier


#### Does this PR introduce a user-facing change?
```release-note
NONE
```

SergeyKanzhelev commented:
/retest

SergeyKanzhelev commented:
/cc @liggitt @mrunalp

liggitt commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -4627,24 +4627,41 @@ type PodStatus struct {
 	// init container will have ready = true, the most recently started container will have
 	// startTime set.
+	// The list must not contain entries with the same container name, and must only contain
+	// statuses for containers listed in pod spec.  However this is not enforced.
```
```suggestion
	// statuses for initContainers listed in pod spec.  However this is not enforced.
```

liggitt commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -4627,24 +4627,41 @@ type PodStatus struct {
+
+	// Statuses for any ephemeral containers that have run in this pod.
+	// The list must not contain entries with the same container name, and must only contain
+	// statuses for containers listed in pod spec. However this is not enforced.
```
```suggestion
	// statuses for ephemeralContainers listed in pod spec. However this is not enforced.
```

SergeyKanzhelev committed:
added a comment that statuses lists are not being validated

SergeyKanzhelev committed:
generated files

seans3 commented:
/triage accepted

alculquicondor commented on pkg/apis/core/types.go:
```diff
@@ -3960,16 +3960,32 @@ type PodStatus struct {
+	// Statuses of init containers in this pod. The most recent successful non-restartable
 	// init container will have ready = true, the most recently started container will have
 	// startTime set.
+	// The list must not contain entries with the same container name, and must only contain
```
If it's not enforced, then the right word is `should`

SergeyKanzhelev replied:
fixed

SergeyKanzhelev commented:
/triage accepted
/cc @haircommander

haircommander commented on api/openapi-spec/v3/api__v1_openapi.json:
```diff
@@ -5696,7 +5696,7 @@
           },
           "containerStatuses": {
-            "description": "The list has one entry per container in the manifest. More info: https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle#pod-and-container-status",
+            "description": "Statuses of containers in this pod. The list should not contain entries with the same container name, and should only contain statuses for containers listed in pod spec. However this is not enforced. If status for not-existing container is present in the list, the behavior of various Kubernetes components is not defined and those statuses might be ignored. More info: h ...
```
`The list should not contain entries with the same container name` is a little clunky to me maybe

`Each container in the pod should have at most one status in this list, and all statuses should be for containers in the pod.`

also maybe `a non-existent` instead of `not-existing`

SergeyKanzhelev replied:
thanks, updated

haircommander commented on api/openapi-spec/v3/api__v1_openapi.json:
```diff
@@ -5807,7 +5807,7 @@
           },
           "containerStatuses": {
-            "description": "The list has one entry per container in the manifest. More info: https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle#pod-and-container-status",
+            "description": "Statuses of containers in this pod. Each container in the pod should have at most one status in this list, and all statuses should be for containers in the pod. However this is not enforced. If status for not-existent container is present in the list, or list have duplicate names, the behavior of various Kubernetes components is not defined and those statuses might be  ...
```
sorry another grammar nit:
```
If a status for a non-existent container is present in the list, or the list has duplicate names,...
```

SergeyKanzhelev replied:
thank you! fixed up

googs1025 commented:
hi, @SergeyKanzhelev 
We can update this code and now solve this flake test problem.

https://github.com/kubernetes/kubernetes/issues/126801

SergeyKanzhelev commented:
/retest

SergeyKanzhelev commented:
/assign @thockin 

@thockin can you please take a look?

SergeyKanzhelev commented:
@soltysh can you please review the printer part of this change? I can split it away into a separate PR if it will be easier

SergeyKanzhelev commented:
@thockin would you be ok taking this one? Or you would prefer if I split the change in two PRs - one with the API text update and one with the printers update?

SergeyKanzhelev commented:
> @thockin would you be ok taking this one? Or you would prefer if I split the change in two PRs - one with the API text update and one with the printers update?

I split it into two PRs

thockin commented:
This seems like a place where ratcheting the validation could apply?  Are we sure we can't fix this?

/lgtm
/approve
/hold on to hope

SergeyKanzhelev commented:
> /hold on to hope

we discussed with @liggitt and didn't find a solution. Not a straightforward one at lease.

Let's merge this doc update and if we can find the way to make it better, we will revise the doc. I am not marking the bug as fixed:

> Related to: https://github.com/kubernetes/kubernetes/issues/124915

/unhold

SergeyKanzhelev commented:
/test pull-kubernetes-e2e-kind-ipv6
